### ISSUE TITLE
confusing arity error

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#768]
```
[list: 1, 3, 2].sort(lam(a, b): a <= b end)
```
produces an error saying the applicant had two arguments but expected one. But above, there is only one argument… [Error resugaring alert.]

**Comment by blerner** [comment]
Explanation: Every method has an extra self argument, so there are indeed _two_ arguments above, namely the list and the lambda.  `sort` doesn't take any comparators as arguments.  `sort-by` takes _two_ comparators (less than, and equality).  I don't think we've workshopped any language surrounding how to describe arity errors about methods, yet.

**Comment by schanzer** [comment]
OK, so this is biting me _hard_ this week with real kids (and these are insanely high performing kids, too!). In an effort to be as correct as possible, I think we're leaking too much sugar/semantics/implementation in the error, to deleterious effect.

My gut instinct is to say that we don't count the `self` argument when we report on the number of inputs. Since methods can't be called on their own anyway (and will generate a different error when they are!), a user will only encounter this error when `self` is supplied. From the user's perspective, the the `sort` method is only going to have one argument. This error basically assumes that the user gets how methods are desugared - and if they don't, it forces them to (a) confront it and understand it or (b) shrug and figure that error messages may not always be useful/sensical.

**Comment by shriram** [comment]
I agree with this. I don't really understand @blerner's response, which only told me what I knew, but didn't actually address the _issue_.

**Comment by schanzer** [comment]
Thank god my kids here are sharp: they grokked the connection between methods and functions quickly. And yet, when chewing on a programming problem, they totally forget — and get confused/annoyed by the artist issue
